enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,402 @@
+---
+title: out-of-tree-provider-support
+authors:
+  - "@Danil-Grigorev"
+  - "@Fedosin"
+reviewers:
+  - "@enxebre"
+  - "@JoelSpeed"
+  - "@crawford"
+  - "@derekwaynecarr"
+  - "@eparis"
+  - "@mrunalp"
+  - "@sttts"
+approvers:
+  - "@enxebre"
+  - "@JoelSpeed"
+  - "@crawford"
+  - "@derekwaynecarr"
+  - "@enxebre"
+  - "@eparis"
+  - "@mrunalp"
+  - "@sttts"
+creation-date: 2020-08-31
+last-updated: 2020-09-30
+status: implementable
+replaces:
+superseded-by:
+---
+
+# Add support for out-of-tree cloud provider integration
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open questions
+
+1. Should the post-install phase of the CCMs be managed by a static-pod, similarly as it is done for the bootstrapping phase, or with a `Deployment`?
+
+Pros and cons:
+
+Deployment:
++ Simple lifecycle management
++ Simple configuration changes
+- Requires additional RBAC configurations and running kubernetes components
+- Requires integration with the cloud credentials operator to access cloud credentials
+
+Static pod:
++ Does not require a running/healthy etcd, api server and kube-controller-manager
++ Thus will be used during bootstrapping phase
+- Requires management with an operator
+- Limited configuration options, and could not be reused from the bootstrapping template, each cloud will require its own template
+
+2. CCM relation to other openshift components, such as SDN and storage? How a non-operational CCM will affect cluster health, which components will take a hit?
+
+- Currently known `kubelet` dependency added requirement to run the component at bootstrapping phase.
+- Storage CSI migration will be required for OpenStack and vSphere providers
+- Networking:
+  - In GCP and OpenStack the `Routes` [interface](https://github.com/kubernetes/cloud-provider/blob/0dc0419ec04907f9deb6aa498cece84f10044726/cloud.go#L63) is implemented. Does it have any relation to the SDN? [Link](https://github.com/kubernetes/cloud-provider/blob/master/controllers/route/route_controller.go) to the controller managing this.
+  - OpenStack Load Balancing is an optional feature, though is [supported](https://github.com/kubernetes/cloud-provider/blob/master/controllers/service/controller.go) by the CCM interface with the [LoadBalancer()](https://github.com/kubernetes/cloud-provider/blob/0dc0419ec04907f9deb6aa498cece84f10044726/cloud.go#L49]. How important is this functionality during cluster runtime?
```

Comment:
So (as I understand it) we require that either the cluster has a CloudProvider that implements LoadBalancer services (which is how most OCP clusters work), or else that you provide some other way to make (OCP) Routes work (which is what we do on bare metal, and I guess some OpenStack installs?). If the cluster is relying on CloudProvider LoadBalancer support, then if the CCM went away, then changes to LoadBalancer Services would not be processed until it came back.

I don't think we try to create any LoadBalancer Services at install time until after the worker nodes are created (at which point cluster-ingress-operator will be trying to create the default router), so as long as CCM is working by then, things should work. I'm not sure if ingress/routes creates new any LoadBalancer Services _after_ install time. If they did, and CCM was unavailable, then the ingress/route would not be usable until CCM came back to set up the load balancer. Of course if the user is creating LoadBalancer Services of their own, then that would also fail if CCM wasn't available.

Maybe need some Network Edge input here

Reply:
From the slack [discussion](https://coreos.slack.com/archives/CCH60A77E/p1601565268029300) @Miciah confirmed this. The `router-default` LB is created for the ingress-operator deployment, so having CCM ready at this point and reporting available condition + no alerts firing guaranties that everything will be working.